syntax = "proto3";

package smart.fridge;

import "google/protobuf/empty.proto";

message Item {
  string name = 1;
  int32 quantity = 2;
}

message FridgeState {
  float fridge_temperature = 1;
  float freezer_temperature = 2;
  bool lightOn = 3;
  repeated Item fridgeContents = 4;
  repeated Item freezerContents = 5;
}

message ContentsResponse {
  repeated Item fridgeContents = 1;
  repeated Item freezerContents = 2;
}

message LightResponse {
  bool lightOn = 1;
}

message TemperatureRequest {
  float targetTemperature = 1;
}

message TemperatureResponse {
  float temperature = 1;
}

message Ack {
  string message = 1;
}

service Fridge {
  rpc GetState (google.protobuf.Empty) returns (FridgeState);

  rpc GetFridgeTemperature (google.protobuf.Empty) returns (TemperatureResponse);
  rpc GetFreezerTemperature (google.protobuf.Empty) returns (TemperatureResponse);

  rpc SetFridgeTemperature (TemperatureRequest) returns (Ack);
  rpc SetFreezerTemperature (TemperatureRequest) returns (Ack);

  rpc GetLightStatus (google.protobuf.Empty) returns (LightResponse);
  rpc ToggleLight (google.protobuf.Empty) returns (Ack);

  rpc GetContents (google.protobuf.Empty) returns (ContentsResponse);

  rpc AddProductToFridge (Item) returns (Ack);
  rpc AddProductToFreezer (Item) returns (Ack);

  rpc RemoveProductFromFridge (Item) returns (Ack);
  rpc RemoveProductFromFreezer (Item) returns (Ack);
}
